src/distributed_join.cuh

```diff
@@ -152,6 +163,162 @@ void all_to_all_comm(cudf::table_view input,
   }
 }
 
+struct compression_functor {
+  template <typename T, std::enable_if_t<std::is_integral<T>::value> * = nullptr>
+  void operator()(const void *uncompressed_data,
+                  size_t uncompressed_count,
+                  rmm::device_buffer &compressed_data,
+                  size_t &compressed_size)
+  {
+    nvcomp::CascadedCompressor<T> compressor(
+      static_cast<const T *>(uncompressed_data), uncompressed_count, 1, 1, true);
+
+    const size_t temp_size = compressor.get_temp_size();
+    rmm::device_buffer temp_space(temp_size);
+    compressed_size = compressor.get_max_output_size(temp_space.data(), temp_size);
+    compressed_data = rmm::device_buffer(compressed_size);
+
+    compressor.compress_async(
+      temp_space.data(), temp_size, compressed_data.data(), &compressed_size, cudaStreamDefault);
+    CUDA_RT_CALL(cudaStreamSynchronize(cudaStreamDefault));
+  }
+
+  template <typename T, std::enable_if_t<!std::is_integral<T>::value> * = nullptr>
+  void operator()(const void *uncompressed_data,
+                  size_t uncompressed_count,
+                  rmm::device_buffer &compressed_data,
+                  size_t &compressed_size)
+  {
+    assert(false && "Unsupported data type for cascaded compressor");
+  }
+};
```

Comment:
Can we just delete this?  Then it should catch the error at compile time instead of at runtime, which is better.

What happens if this code is removed?

Reply:
I used `enable_if_t` because my understanding is that cascaded compression only works with integral types. It will not generate a compile error if removed, I'm not sure what nvcomp will do if we pass a generic cuDF type to it.

cc @nsakharnykh for more inputs on nvcomp side.

Reply:
If you're able to remove this without causing a compile error, that means that it is generating no code.  c++ can't make an instance of a template at run-time so it's not possible for this to make a difference at runtime, either.  So it seems like this code isn't needed at compile time and can't be used at runtime.  So it can never be used.

Might as well delete it.  If you're really worried, just comment it out.

If you leave it in, you will potentially cause runtime errors that could have been caught earlier at compile time.  It's also extra code to maintain that is untested.  We don't even know if it does what it's meant to do because it has never been used.

Reply:
> If you're able to remove this without causing a compile error, that means that it is generating no code. c++ can't make an instance of a template at run-time so it's not possible for this to make a difference at runtime, either. So it seems like this code isn't needed at compile time and can't be used at runtime. So it can never be used.

No compile error in one path does not mean a separate path is not used. It will generate runtime error in the other path. 

> If you leave it in, you will potentially cause runtime errors that could have been caught earlier at compile time. It's also extra code to maintain that is untested. We don't even know if it does what it's meant to do because it has never been used.

I tentatively disagree. We don't know whether nvcomp is compatible with all possible cuDF types. Let's say, if cuDF adds a new data type in the future that is incompatible but nvcomp produces no compile error, it could lead the program to finish without error but the sematic is wrong. I would prefer a runtime error than no error. I think it's probably the best if we "opt-in", as in only supporting the types we know that are compatible. Plus, in the best case, if it is indeed a compile error, we need to add this back.

Reply:
> No compile error in one path does not mean a separate path is not used. It will generate runtime error in the other path.

I don't understand this part.  If you're able to compile without this code then the code is unused and it can never be used.  So how could you get a runtime error for a function that doesn't exist?

I don't understand how you can have a situation where, with this code, there is a runtime error, but without this code, there is no compile error and no runtime error.  That's impossible with the way that the template is written now, I think.

The only way to reach this runtime error is if the source is generating machine code.  But if you're able to compile without this code, then there is nothing being generated.  So how would you ever execute a function that doesn't even assemble?

Maybe I'm just confused about it but if you're able to compile without this then I think that you should.

Reply:
Oh I think I misunderstood your original question. You asked if we remove this part but keeps the `enable_if_t` in the other path, whether this code will compile. Is my understanding correct this time? The answer is it will not compile.

We can remove `enable_if_t` in both path, and throw it to nvcomp. Then it will compile but I think that's dangerous.

Reply:
I think `type_dispatcher` will instantiate all possible cuDF types for this functor, and then, if we don't have a special treatment for types not supported by nvcomp, nvcomp will throw an exception at runtime: https://github.com/NVIDIA/nvcomp/blob/64acf248ba954152aba1f51fc938ffe015ef1e46/include/nvcomp.hpp#L307. I'm curious if it would even compile, @gaohao95 can you try it and post here if you get a complication error? I'd like to know which type doesn't compile for nvcomp (if any), so we can track it.

Reply:
If we don't have `enable_if_t` at all in both path, all cuDF types will be passed to nvcomp, and it will compile just fine.

Reply:
ok, thanks for confirming! I believe we can remove it then, and just rely on nvcomp reporting the errors at runtime. nvcomp may support other types in the future for cascaded (and it should already support that for lz4), so it's better to avoid any custom logic for this in distributed join. What do you think?

Reply:
Sure, since nvcomp has this check in place, I'm okay with not checking the type compatibility in distributed join.